vending-app/src/main/kotlin/com/google/android/finsky/splitinstallservice/SplitInstallManager.kt

```diff
@@ -0,0 +1,337 @@
+            }
+        }
+        Log.d(TAG, "requestDownloadUrls end packSet -> $packSet")
+        return packSet.onEach { updateSplitInstallRecord(callingPackage, it) }
+    }
+
+    private suspend fun getOauthToken(): String {
+        val accounts = AccountManager.get(context).getAccountsByType(DEFAULT_ACCOUNT_TYPE)
+        var oauthToken: String? = null
+        if (accounts.isEmpty()) {
+            Log.w(TAG, "No Google account found")
+            throw RuntimeException("No Google account found")
+        } else for (account: Account in accounts) {
+            oauthToken = try {
+                getAuthToken(AccountManager.get(context), account, AUTH_TOKEN_SCOPE).getString(AccountManager.KEY_AUTHTOKEN)
+            } catch (e: AuthenticatorException) {
+                Log.w(TAG, "Could not fetch auth token for account $account")
+                null
+            }
+            if (oauthToken != null) {
+                break
+            }
+        }
+        return oauthToken ?: throw RuntimeException("oauthToken is null")
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun checkSplitInstalled(callingPackage: String, splitName: String): Boolean {
+        if (!splitInstallRecord.containsKey(callingPackage)) return false
+        return splitInstallRecord[callingPackage]?.find { it.first == splitName }?.third != STATUS_UNKNOWN
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun updateSplitInstallRecord(callingPackage: String, triple: Triple<String, String, Int>) {
+        splitInstallRecord[callingPackage]?.let { triples ->
+            val find = triples.find { it.first == triple.first }
+            find?.let { triples.remove(it) }
+            triples.add(triple)
+        } ?: run {
+            val triples = ArraySet<Triple<String, String, Int>>()
+            triples.add(triple)
+            splitInstallRecord[callingPackage] = triples
+        }
+    }
+
+    private fun notify(context: Context) {
+        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            notificationManager.createNotificationChannel(
+                NotificationChannel(NOTIFY_CHANNEL_ID, NOTIFY_CHANNEL_NAME, NotificationManager.IMPORTANCE_DEFAULT)
+            )
+        }
+        NotificationCompat.Builder(context, NOTIFY_CHANNEL_ID).setSmallIcon(android.R.drawable.stat_sys_download)
+            .setContentTitle(context.getString(R.string.split_install, context.getString(R.string.app_name))).setPriority(NotificationCompat.PRIORITY_DEFAULT)
+            .setDefaults(
+                NotificationCompat.DEFAULT_ALL
+            ).build().also {
+                notificationManager.notify(SPLIT_INSTALL_NOTIFY_ID, it)
+            }
+    }
+
+    private fun Context.splitSaveFile() = File(filesDir, FILE_SAVE_PATH)
```

Comment:
Installed packages are deleted after installation, but if installation fails and never retried (for instance: one large package is downloaded successfully, then a small package is downloaded unsuccessfully due to a network error, then the user uninstalls the app because it doesn't work), this can still lead to storage leaks from downloaded-but-never-installed apps. We should use `cacheDir` instead: it will allow the system and users to clear the cache themselves.